p2p_beef_tx_test.go

```diff
@@ -0,0 +1,227 @@
+			expectedStream: "",
+			expectedError:  errors.New("invalid beef marker"),
+		},
+		{
+			name:           "No version encoded",
+			hexString:      "beef01020101cd73c0c6bb645581816fa960fd2f1636062fcbf23cb57981074ab8d708a76e3b02003470d882cf556a4b943639eba15dc795dffdbebdc98b9a98e3637fda96e3811e01c58e40f22b9e9fcd05a09689a9b19e6e62dbfd3335c5253d09a7a7cd755d9a3c04008c00bb9360e93fb822c84b2e579fa4ce75c8378ae87f67730a49552f73c56ee801da256f78ae0ad74bbf539662cdb9122aa02ba9a9d883f1d52468d96290515adb02b4c8d919190a090e77b73ffcd52b85babaaeeb62da000473102aca7f070facef03e5b331f4961d764373f3a4e2751954e75489fb17902aad583eedbb41dc165a3b02020000000158cb8b052fded9a6c450c4212562df8820359ec34da41286421e0d0f2b7eefee000000006a47304402206b1255cb23454c63b22833de25a3a3ecbdb8d8645ad129d3269cdddf10b2ec98022034cadf46e5bfecc38940e5497ddf5fa9aeb37ff5ec3fe8e21b19cbb64a45ec324121029a82bfce319faccc34095c8405896e1223921917501a4f736a04f126d6a01c12ffffffff0101000000000000001976a914d866ec5ebb0f4e3840351ee61887101e5407562988ac000000000100020000000158cb8b052fded9a6c450c4212562df8820359ec34da41286421e0d0f2b7eefee000000006a47304402206b1255cb23454c63b22833de25a3a3ecbdb8d8645ad129d3269cdddf10b2ec98022034cadf46e5bfecc38940e5497ddf5fa9aeb37ff5ec3fe8e21b19cbb64a45ec324121029a82bfce319faccc34095c8405896e1223921917501a4f736a04f126d6a01c12ffffffff0101000000000000001976a914d866ec5ebb0f4e3840351ee61887101e5407562988ac0000000000",
+			expectedStream: "",
+			expectedError:  errors.New("invalid beef marker"),
+		},
+		{
+			name:           "To short hexString",
+			hexString:      "001",
+			expectedStream: "",
+			expectedError:  errors.New("invalid beef hex stream"),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// given
+			stream := tc.hexString
+
+			// when
+			result, err := extractBytesWithoutVersionAndMarker(stream)
```

Comment:
## doubt

I'm not sure if we should have a tests for that function.
I don't mean it's bad that you wrote it, because probably it is very helpful when developing the solution.
But it doesn't mean that we should keep it.
Reason:
1. it is a private, helper function
2. it is not hard to reason and understand algorithm
3. keeping such tests will make it harder (more work to do) when extending a current solution (for example handling of different versions) and when doing a refactoring (for example inlining this function first to split it in separated parts connected with other parts of code)
4. one important thing is that all the cases for that function should be tested at the public function level also, so this force you to make a multiple duplications of the same scenarios, but on a different function levels.

Reply:
I'm not sure I understand you correctly. So you're saying we shouldn't test private functions at all?
I've always thought that breaking down implementation into smaller functions was not only to help with reading, but also to help with testing.

It's usually difficult to test every case from the "public function level", but it's easier to break the code into smaller functions and then test them.
I believe that tests should be read in context and not as separate functions

Reply:
> So you're saying we shouldn't test private functions at all?
Not exactly what I'm saying, but in most cases, this it true statement.
IMHO actually breaking code into smaller pieces is meant to help reading the code, and maintain it. It shouldn't be made just to help testing. If you have problem with testing public functions then probably design of the public function should be changed.
Beside that if you test only private function for some case, what actually prevents the outer function to call it or use it wrongly in that case, for example: not handle this specific error, or errors at all or assuming that returned value exists, but actually in that corner case it is nil, etc.
Beside that, tests should make easy refactoring of code, if you have tests for private functions it makes harder to refactor the code, because you cannot easily play around with this function, change it few times to meet requirements of readable code and outer function logic, but you need to also always adjust the tests, and actually when there are a lot of tests for private methods that could to discourage anyone from making a refactoring.

One thing, during implementation phase you can test private method, to help you implement it correctly, and I found it often very useful in that phase. But we shouldn't keep those tests as a safety net for the code and future changes.

Reply:
Ok, I decided to mark all methods/functions unexported but DecodeBEEF. It means that only this function is currently covered with unit tests.